velox/functions/sparksql/AtLeastNNonNulls.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct AtLeastNNonNullsFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void callNullable(
+      out_type<bool>& out,
+      const arg_type<int32_t>* n,
+      const arg_type<Variadic<Any>>* inputs) {
+    out = getNonNullAndNanCount(inputs) >= *n;
+  }
+
+ private:
+  int32_t getNonNullAndNanCount(const arg_type<Variadic<Any>>* inputs) {
+    int32_t result = 0;
+    for (const auto& input : *inputs) {
```

Comment:
We should check if can end up for loop early, https://github.com/apache/spark/blob/110b558570176d5d5ee5ba85bb071fd66a94a7a0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L437

Check the nonNullAndNanCount > n, if it is wide columns table, it will benefit.

Reply:
Good catch, thanks.